src/llama_recipes/inference/llm.py

```diff
@@ -0,0 +1,197 @@
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+from __future__ import annotations
+
+import logging
+
+import time
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import openai
+from langchain_together import Together
+
+from typing_extensions import override
+
+
+NUM_LLM_RETRIES = 10
+
+MAX_TOKENS = 1000
+
+LOG: logging.Logger = logging.getLogger(__name__)
+
+
+class LLM(ABC):
+    def __init__(self, model: str, api_key: str | None = None) -> None:
+        if model not in self.valid_models():
+            LOG.warning(
+                f"{model} is not in the valid model list for {type(self).__name__}. Valid models are: {', '.join(self.valid_models())}."
+            )
+        self.model: str = model
+        self.api_key: str | None = api_key
+
+    @abstractmethod
+    def query(self, prompt: str) -> str:
+        """
+        Abstract method to query an LLM with a given prompt and return the response.
+
+        Args:
+            prompt (str): The prompt to send to the LLM.
+
+        Returns:
+            str: The response from the LLM.
+        """
+        pass
+
+    def query_with_system_prompt(self, system_prompt: str, prompt: str) -> str:
+        """
+        Abstract method to query an LLM with a given prompt and system prompt and return the response.
+
+        Args:
+            system prompt (str): The system prompt to send to the LLM.
+            prompt (str): The prompt to send to the LLM.
+
+        Returns:
+            str: The response from the LLM.
+        """
+        return self.query(system_prompt + "\n" + prompt)
+
+    def _query_with_retries(
+        self,
+        func: Callable[..., str],
+        *args: str,
+        retries: int = NUM_LLM_RETRIES,
+        backoff_factor: float = 0.5,
+    ) -> str:
+        last_exception = None
+        for retry in range(retries):
+            try:
+                return func(*args)
+            except Exception as exception:
+                last_exception = exception
+                sleep_time = backoff_factor * (2**retry)
+                time.sleep(sleep_time)
+                LOG.debug(
+                    f"LLM Query failed with error: {exception}. Sleeping for {sleep_time} seconds..."
+                )
+        raise RuntimeError(
+            f"Unable to query LLM after {retries} retries: {last_exception}"
+        )
+
+    def query_with_retries(self, prompt: str) -> str:
+        return self._query_with_retries(self.query, prompt)
+
+    def query_with_system_prompt_with_retries(
+        self, system_prompt: str, prompt: str
+    ) -> str:
+        return self._query_with_retries(
+            self.query_with_system_prompt, system_prompt, prompt
+        )
+
+    def valid_models(self) -> list[str]:
+        """List of valid model parameters, e.g. 'gpt-3.5-turbo' for GPT"""
+        return []
+
+
+class OPENAI(LLM):
```

Comment:
Also, wondering if we need to keep the specific APIs out of this module and keep them as examples? otherwise it means one need to install all these packages/ adding all as dependency? openai, ocotoAI etc?

Reply:
I have deliberately left these modules out of the requirements.txt to avoid having a module level dependency on these modules. I think that users can install them selectively as they need to use them and get an error when they are not found. Do you think that's a reasonable middle ground?